### PR TITLE
Don't suggest unnameable generic arguments

### LINKED ISSUE #99565: “consider specifying the generic arguments” is confusing when some types are unnameable
Given the following code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0a9a81dc70d5d9e9873ad7b65a6975d9

```rust
fn foo<T, U>(_: U) {}

fn bar() {
    foo(|| {});
}
```

The current output is:

```
error[[E0282]](https://doc.rust-lang.org/stable/error-index.html#E0282): type annotations needed
 --> src/lib.rs:4:5
  |
4 |     foo(|| {});
  |     ^^^ cannot infer type of the type parameter `T` declared on the function `foo`
  |
help: consider specifying the generic arguments
  |
4 |     foo::<T, [closure@src/lib.rs:4:9: 4:11]>(|| {});
  |        +++++++++++++++++++++++++++++++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground` due to previous error
```

Ideally the output should look like:

```
error[[E0282]](https://doc.rust-lang.org/stable/error-index.html#E0282): type annotations needed
 --> src/lib.rs:4:5
  |
4 |     foo(|| {});
  |     ^^^ cannot infer type of the type parameter `T` declared on the function `foo`
  |
help: consider specifying the generic arguments
  |
4 |     foo::<T, _>(|| {});
  |        ++++++++

For more information about this error, try `rustc --explain E0282`.
error: could not compile `playground` due to previous error
```

Or alternatively shouldn’t contain the suggestion at all.

### PR DESCRIPTION
**Pull request by fmease**
Fixes #99565.

@rustbot label T-compiler A-diagnostics
r? @rust-lang/wg-diagnostics

### CODE REVIEW

File: compiler/rustc_middle/src/ty/diagnostics.rs [hunk 1]
```diff
@@ -426,6 +426,8 @@ impl<'tcx> TypeVisitor<'tcx> for IsSuggestableVisitor<'tcx> {
 
     fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
         match t.kind() {
+            Infer(InferTy::TyVar(_)) => {}
```
Comment (fmease) [inline compiler/rustc_middle/src/ty/diagnostics.rs:429]: I doubt that this is the correct approach.

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 2]
```diff
@@ -407,11 +408,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
+                        for arg in generic_args {
+                            printer.write_str(", ")?;
+                            printer = print(arg, printer)?;
+                        }
+                    }
+
+                    printer.into_buffer()
+                };
+                let args = args.unwrap();
+
```
Comment (fmease) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:437]: Unfortunately, I couldn't use `printer.comma_sep` anymore being too inflexible and I had to inline it.

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 3]
```diff
@@ -407,11 +408,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
                 err.span_suggestion_verbose(
                     insert_span,
-                    &format!("consider specifying the generic argument{}", pluralize!(args.len()),),
+                    &format!(
+                        "consider specifying the generic argument{}",
+                        pluralize!(generic_args.len()),
```
Comment (fmease) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:442]: `args.len()` was incorrect: `args` is a `String` here, not a vector of arguments. It was measuring the amount of bytes in the string, not the number of arguments.

**Comment by fmease** [comment]
Several minor diagnostic regressions. Not sure how to fix them.

File: compiler/rustc_middle/src/ty/diagnostics.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/diagnostics.rs:429]: This is _almost_ the right approach -- instead, can you make it a boolean? Like `infer_suggestable`? And edit all of the `is_suggestable` calls to be false except for the one you added?

File: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs [hunk 4]
```diff
@@ -407,11 +408,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
                 err.span_label(span, cannot_infer_msg);
 
-                let printer = fmt_printer(self, Namespace::TypeNS);
-                let args = printer.comma_sep(generic_args.iter().copied()).unwrap().into_buffer();
+                let args: Result<_, std::fmt::Error> = try {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:411]: Can you instead preprocess `generic_args` first -- replacing args that are not suggestible with a fresh ty or const var? Then pass it to comma_sep.

**Comment by fmease** [comment]
I applied your feedback. It resolved those diagnostic regressions which is great!

**Pull request merged**